### PR TITLE
Disallow declaration of variables within constructor/field initializers.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Related to #12597.

@dotnet/roslyn-compiler, @gafter Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs [hunk 1]
```diff
@@ -5012,11 +5193,22 @@ public C(object b)
 }
 ";
             var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe);
+#if ALLOW_IN_CONSTRUCTOR_INITIALIZER
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: I didn't understand the purpose of the `ALLOW_IN_CONSTRUCTOR_INITIALIZER` pragma. Why not just update the test code?

**Comment by jcouv** [comment]
LGTM but I would not introduce #if blocks.

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: The purpose is to keep expectations, we are planning to allow declarations in these contexts in the future.

Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: Makes sense. Thanks

**Comment by AlekseyTs** [comment]
Ping. @dotnet/roslyn-compiler, @gafter Please review.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 2]
```diff
@@ -4947,4 +4947,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_OutVarDeconstructionIsNotSupported" xml:space="preserve">
     <value>Deconstruction is not supported for an 'out' argument.</value>
   </data>
+  <data name="ERR_ExpressionVariableInConstructorOrFieldInitializer" xml:space="preserve">
+    <value>Out variable or pattern variable declarations are not allowed within constructor/field/auto-implemented property initializers.</value>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Consider separating cases with comma for consistency with other errors. And remove `auto-implemented` if not necessary. Perhaps: `... not allowed within constructors, field initializers, or property initializers`.

Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Also, I believe the first "or" should be "and" (they are *both* forbidden), so I suggest

> Out variable and pattern variable declarations are not allowed within constructor initializers, field initializers, or property initializers.

This is long-winded enough that you could consider specializing it.

---
In reply to: [77212251](https://github.com/dotnet/roslyn/pull/13510#discussion_r77212251) [](ancestors = 77212251)

**Comment by gafter** [comment]
Other than a slight wording change on the diagnostic, LGTM

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Opened #13528.

**Pull request merged**